src/app/ui/theme-context.tsx

```diff
@@ -0,0 +1,77 @@
+'use client'
+import React from "react"
+import { getThemePreference, setThemePreference } from "./dark-mode"
+import { colors } from "./colors"
+import { Box } from "./components"
+
+export type Theme = 'light' | 'dark'
+
+const ThemeContext = React.createContext<{
+    theme: Theme | undefined,
+    preference: Theme | undefined,
+    setTheme: (theme: Theme | undefined) => void
+}>({ theme: undefined, preference: undefined, setTheme: (colorMode) => { } })
+
+export const useTheme = () => React.useContext(ThemeContext)
+
+const getSystemTheme = () => {
+    return window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light'
+}
+
+export const ThemeProvider = ({ children }: { children: React.ReactNode }) => {
+    const [theme, rawSetTheme] = React.useState<Theme | undefined>(undefined)
+    const [preference, setPreference] = React.useState<Theme | undefined>(undefined)
+
+    React.useEffect(() => {
+        const root = window.document.documentElement
+        const initialMode = getComputedStyle(root).getPropertyValue('--initial-color-mode')
+        // initialMode should always be light or dark, using the preference if saved or the system default.
+        rawSetTheme(initialMode === 'light' ? 'light' : 'dark')
+        // getPreference may return return light or dark if one has been set, or undefined if system default.
+        setPreference(getThemePreference())
+    }, [])
+
+    React.useEffect(() => {
+        const mql = window.matchMedia('(prefers-color-scheme: dark)')
+        mql.onchange = (e) => {
```

Comment:
at first I did this with `addEventListener`, however I was having trouble getting the event listeners removed to clean up, which was causing all sorts of bugs. I spent a while investigating how maybe the function reference was getting lost, but was taking too long and found this solution that works in the mean time.